
## callback

> 把a函数传递到b函数中，并且b函数执行a函数，就称a函数为回调函数，如果没有名称，就称作匿名回调函数

### 缺点：
- 嵌套，可读性很差

## promise

> js是单线程执行的，所以部分的操作必须是异步执行，在异步执行中通常使用回调函数来实现，但是回调函数的写法不好看而且不利于复用，才出现了promise对象  

```javascript

let p=new Promise(fn)
p.then()
p.catch()

```
### 优点：
 - 把执行代码和结果处理代码清晰分离，避免多任务串行时的嵌套
 - 支持多个任务并行执行`Promise.all([j1,j2])`
 - 支持容错的同时向多个任务请求，只需获得先返回的结果`Promise.race([j1,j2])` 

### 缺点：
 - 无法取消
 - 不设置回调函数，抛出的错误不会反应到外部
 - pending状态时，无法得知进展

### promise的实现

```javascript

/*
我们要满足状态只能三种状态：PENDING,FULFILLED,REJECTED三种状态，且状态只能由PENDING=>FULFILLED,或者PENDING=>REJECTED
*/
var PENDING = 0;
var FULFILLED = 1;
var REJECTED = 2;
/*
value状态为执行成功事件的入参，deferreds保存着状态改变之后的需要处理的函数以及promise子节点，构造函数里面应该包含这三个属性的初始化
    */
function Promise(callback) {
    this.status = PENDING;
    this.value = null;
    this.defferd = [];
    setTimeout(callback.bind(this, this.resolve.bind(this), this.reject.bind(this)), 0);
}

Promise.prototype = {
    constructor: Promise,
    //触发改变promise状态到FULFILLED
    resolve: function (result) {
        this.status = FULFILLED;
        this.value = result;
        this.done();
    },
    //触发改变promise状态到REJECTED
    reject: function (error) {
        this.status = REJECTED;
        this.value = error;
    },
    //处理defferd
    handle: function (fn) {
        if (!fn) {
            return;
        }
        var value = this.value;
        var t = this.status;
        var p;
        if (t == PENDING) {
                this.defferd.push(fn);
        } else {
            if (t == FULFILLED && typeof fn.onfulfiled == 'function') {
                p = fn.onfulfiled(value);
            }
            if (t == REJECTED && typeof fn.onrejected == 'function') {
                p = fn.onrejected(value);
            }
        var promise = fn.promise;
        if (promise) {
            if (p && p.constructor == Promise) {
                p.defferd = promise.defferd;
            } else {
                p = this;
                p.defferd = promise.defferd;
                this.done();
            }
        }
        }
    },
    //触发promise defferd里面需要执行的函数
    done: function () {
        var status = this.status;
        if (status == PENDING) {
            return;
        }
        var defferd = this.defferd;
        for (var i = 0; i < defferd.length; i++) {
            this.handle(defferd[i]);
        }
    },
    /*储存then函数里面的事件
    返回promise对象
    defferd函数当前promise对象里面
    重点：then中返回的是一个新的promise、等待这个promise的运行结束，才调用resolve去handler
    */
    then: function (success, fail) {
        var o = {
            onfulfiled: success,
            onrejected: fail
        };
        var status = this.status;
        o.promise = new this.constructor(function () {

        });
        if (status == PENDING) {
            this.defferd.push(o);
        } else if (status == FULFILLED || status == REJECTED) {
            this.handle(o);
        }
        return o.promise;
    }
};


```
> 参考资:https://segmentfault.com/a/1190000006103601

### promise与setTimeout执行顺序

> macro-task(宏任务)：setTimeout,setInterval
> micro-task(微任务)：Promise,process.nextTick

```javascript
//宏任务，放到event queue里面
setTimeout(function() {
    console.log('setTimeout');
})
//直接执行
new Promise(function(resolve) {
    console.log('promise');
}).then(function() {
    //放到微任务
    console.log('then');
})
//直接执行
console.log('console');

//promise
//console
//then 先执行微任务
//settimeout 执行完微任务，第一轮结束，进行第二轮，先找刚刚的宏任务执行

```

## async/await
> 函数的修饰符，async会隐式的返回一个promise，函数的返回值将作为promise resolve的值，await后面必须跟着一个promise，await只能出现在async函数内
### 什么是async

```javascript
// async是Generator函数的语法糖

// Generator版本
function* (){
    let f1=yield p1();
    f1.xx()
}

// async
async function(){
    let f1=await p1();
    f1.xx()
}
```
### promise的几个不足
- 仍然不够简洁，仍然需要创建then调用链，把返回值一层一层传递给下一个then
- 异步不会向上抛出。要在then中才会抛出，在then外写try，catch是没有效果的
- 调试打断点时，会跳过异步代码，导致找不到错误

### async/await的优点
- 语义更清楚。
- 代码简洁明了，易于阅读和理解，不需要写.then，不需要写匿名函数处理的resolve值，也不需要定义多余的data变量
- 抛出的异常可以被try/catch到
- 条件语句会导致在promise then中多层嵌套，async/await就可以提高可读性
- 调试时可以像调试同步代码一样，一个个await跳过

### 注意
- await后面的promise如果被reject，整个async函数都会中断执行，如果不希望后面的操作被中断，可以放在try/catch里面


